core-backend/sandbox/src/funcs.rs

```diff
@@ -92,23 +93,27 @@ pub enum FuncError<E> {
     DebugString(FromUtf8Error),
     #[display(fmt = "`gr_error` expects error occurred earlier")]
     SyscallErrorExpected,
-    #[display(fmt = "`gr_exit` has been called")]
-    Exit,
-    #[display(fmt = "`gr_leave` has been called")]
-    Leave,
-    #[display(fmt = "`gr_wait` has been called")]
-    Wait,
-    #[display(fmt = "Unable to call a forbidden function")]
-    ForbiddenFunction,
+    #[display(fmt = "Terminated: {:?}", _0)]
+    Terminated(TerminationReason),
 }
 
-impl<E> FuncError<E> {
+impl<E> FuncError<E>
+where
+    E: fmt::Display,
+{
     fn as_core(&self) -> Option<&E> {
         match self {
             Self::Core(err) => Some(err),
             _ => None,
         }
     }
+
+    pub fn into_termination_reason(self) -> TerminationReason {
+        match self {
+            Self::Terminated(reason) => reason,
+            err => TerminationReason::Trap(TrapExplanation::Other(err.to_string())),
```

Comment:
As I said previously, can't we pass here not string, but the cow? For future encoding, we don't need to allocate always.

Reply:
The lifecycle of errors would be
created -> thrown to upper layer -> (optionally) printed in `log::debug` -> encoded via `(E: Encode) trap.encode()` -> sent as bytes after encode.

`&'static str` encodes the same as `String`, but need to check it if I'm wrong, so no need in allocating string always .

Reply:
I wondering do we actually need it now as `Other` variant is always built with `String` and is it premature optimisations or not

Reply:
Ok, lets left it as it is atm